Add support for tooltips

doug opened the pull request:
Please:
- [x] Make your pull request atomic, fixing one issue at a time unless there are many relevant issues that cannot be decoupled.
- [x] Provide a test case & update the documentation under `site/docs/`
- [x] Make lint and test pass. (Run `npm run lint` and `npm run test`)
- [x] Make sure you have merged `master` into your branch. 
- [x] Provide a concise title so that we can just copy it to our release note. 
  - Use imperative mood and present tense.
  - Mention relevant issues. (e.g., `#1`)

domoritz commented:
Thank your for the pull request @doug! Before we review this, note that we are working on a separate library for flexible tooltips https://github.com/vega/vega-tooltip. However, a separate tooltip channel could be great to define what should be shown.

doug commented:
Makes sense. A more flexible tooltip would be great. This is clearly limited, as it is only using the tooltip string property from the Vega spec and passing it through. Still useful as I calculate a tooltip string and use that field for the tooltip. Let me know if there is anything I can do to help. I clearly want this to happen :) Will the vega-tooltip library still be declarative and how will it relate to the spec? I still want to be able to serialize the entire representation including tooltips without having to call javascript.

kanitw commented:
@doug Thanks so much for your PR.  

I agree that this is definitely useful but we're a bit busy this week so there might be some delay for reviewing.  I hope to review this PR this weekend or next week.  

The tooltip library will probably require an additional call to JavaScript, but can produce a more powerful output than what standard `tooltip` channel in purely Vega could do.

doug commented:
No worries, understand being busy. And no rush on this. Honestly, I wasn't quite sure what the best place was in the code to extend, so let me know if this is useful at all, and feel free to ignore. :)

I understand tooltips in javascript are very powerful (a lot more control with less verbosity), but I would like to push back a little bit and express how much I would prefer to be able to express them declaratively in the vega lite spec.

Clearly nice tooltips can be made in Vega declaratively (https://github.com/vega/vega/blob/master/test/specs-valid/budget-forecasts.vg.json) using groups. 

For me I also care about expressing and generating the spec from languages other than javascript (particularly python at the moment) so having to call additional javascript is something I want to avoid.

So even without compiling into some complex group just using the very basic tooltip string field in Vega does most of what I need.

This patch basically allow for:
```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
  "description": "A simple bar chart with embedded data.",
  "data": {
    "values": [
      {"a": "A","b": 28}, {"a": "B","b": 55}, {"a": "C","b": 43},
      {"a": "D","b": 91}, {"a": "E","b": 81}, {"a": "F","b": 53},
      {"a": "G","b": 19}, {"a": "H","b": 87}, {"a": "I","b": 52}
    ]
  },
  "transform": [
    {"calculate": "datum.a + ', ' + datum.b", "as": "tooltip"},
  ],
  "mark": "bar",
  "encoding": {
    "x": {"field": "a", "type": "ordinal"},
    "y": {"field": "b", "type": "quantitative"}
    "tooltip": { "field": "tooltip" },
    },
  }
}
```

If there was a desire for more complex declarative tooltips, maybe building something similar to the new "layer" call could allow you to construct a layered encoding inside a tooltip, which would then compile in Vega to a group with multiple marks positioned relative to the mark they are attached to nearest the mouse.

And of course if you want something complex and dynamic dropping to javascript and using the vega-tooltip seems like the way to go.

craigcitro commented:
👍 from the peanut gallery: it would be great to have a lightweight way to do basic tooltips in vega-lite.

domoritz commented:
CC @sirahd

domoritz commented:
I'm trying to understand how this actually shows any tooltips. Can you explain this a bit more @doug?

kanitw commented:
@domoritz  He is attemping to use Vega's new tooltip channel, which output `title` tag in svg.  (Can't remember what Jeff uses for Canvas though.)

domoritz commented:
Hmm, I tried it with Vega 30 and didn't see anything in the SVG output. If it works, we should get this in as a basic tooltip support. vega-tooltip could then read that property. cc @sirahd

kanitw commented:
@domoritz Vega-tooltip could read that property if exists, but it doesn't the default tooltip that vega-tooltip provides is actually way more powerful.

doug commented:
Yes. Just passing the tooltip encoding string property to vega.

A major constraint being that it must be a string and you get no further
control on how it is displayed. Does work with both canvas and svg when I
tested it.

https://vega.github.io/vega/docs/marks/shape/
Tooltip at the bottom of general mark properties.

On Sat, Apr 29, 2017 at 05:31 Kanit Wongsuphasawat <notifications@github.com>
wrote:

> @domoritz <https://github.com/domoritz> Vega-tooltip could read that
> property if exists, but it doesn't the default tooltip that vega-tooltip
> provides is actually way more powerful.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vega/vega-lite/pull/2132#issuecomment-298146458>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAANMyl6hvTY9tFZ4OcXusT8YH9E6ud2ks5r0r0ggaJpZM4M84v4>
> .
>

domoritz commented:
<img width="260" alt="screen shot 2017-04-29 at 09 24 39" src="https://cloud.githubusercontent.com/assets/589034/25557044/cbeb4a62-2cbd-11e7-8de3-02a718950e40.png">

domoritz commented:
This is a great addition @doug! I'm happy to merge after you rebase the commit on the latest master and add a few unit tests. Now that we have more time, we will review your changes promptly!

kanitw commented on src/compile/mark/valueref.ts:
```diff
@@ -135,6 +135,16 @@ export function midPoint(channel: Channel, channelDef: ChannelDef, scaleName: st
   return defaultRef;
 }
 
+export function tooltip(def: FieldDef | ValueDef<string>, config: Config): VgValueRef {
```
This new method doesn't deal with `timeUnit`, `bin`, and `format` correctly yet. 

Instead of creating new helper method here, I believe you can modify `text()` in `mixins.ts`  to accept a second optional parameter 

```
export function text(model: UnitModel, vgChannel: 'text' | 'tooltip' = 'text') {
   ...
}
```

You can then replace `...mixins.tooltip(model)` with `...mixins.text(model, 'text')` in all mark compilers.

With these minor modifications, I believe it should work correctly without a need for new methods in both `valueref.ts` and `mixins.ts`.  (Of course, please try to test some of these different cases.)

doug replied:
Sounds good. Done.

kanitw commented on src/encoding.ts:
```diff
@@ -75,6 +75,11 @@ export interface Encoding {
+  /**
+   * Tooltip adds basic string tooltip support.
+   */
+  tooltip?: FieldDef | ValueDef<string>;
```
This could be `TextFieldDef | ConditionalValueDef<string|number|boolean>` just like text.  

(Since tooltip will be pretty similar to `text`, you might wanna move it to above `order`, right after `text`.)

doug replied:
I think this can only be TextFieldDef<F> | ConditionalValueDef<string>
also moved above order.

kanitw commented on src/compile/mark/mixins.ts:
```diff
@@ -87,6 +87,12 @@ export function text(model: UnitModel) {
   return wrapCondition(model, channelDef && channelDef.condition, 'text', ref.text(channelDef, model.config));
 }
 
+export function tooltip(model: UnitModel) {
```
See my comment in `valueref.ts`.

doug replied:
Done.

kanitw commented on src/encoding.ts:
```diff
@@ -75,6 +75,11 @@ export interface Encoding {
   order?: OrderFieldDef | OrderFieldDef[];
+
+  /**
+   * Tooltip adds basic string tooltip support.
```
"The tooltip text to show upon mouse hover."  would be a bit better.  

(FYI, we use this JSDocs comment in the JSON schema, which is in turn used to populate the docs.)

doug replied:
Done, thanks.

doug commented on src/encoding.ts:
```diff
@@ -114,6 +119,12 @@ export function normalizeEncoding(encoding: Encoding, mark: Mark): Encoding {
       return normalizedEncoding;
     }
 
+    if (channel === 'tooltip') {
```
Removed. I had this in here because normalizeEncoding is called and the normalization of the tooltip tries to add a type, but a type is irrelevant for a tooltip since it has to be a string. This is no longer logging an error with the improved valueref code, so removed.

doug commented:
@kanitw @domoritz it helps to work with a well thought out code base, thanks for all the hard work on the project! I updated the pull request with the modifications (thank you for the suggestions and review @kanitw), all tests are passing, and I added some tests for tooltip as well as updated the docs. Should be good to go. :smile:

domoritz commented:
Awesome! Can you rerun schema to make travis pass?

doug commented:
Schema updated, thanks.

kanitw commented:
@doug Thanks so much.  Your change is now merged into master!

Are you still at Google in Seattle?  If so, we should have lunch/dinner sometimes.

domoritz commented:
Thank you so much for your contribution @doug!

g3o2 commented:
Unfortunately, the spec shown here above does not work with the current implementation (vega-lite 2.0.0-beta.2 and vega 3.0.0-beta.30).
